### PR TITLE
Fix to allow etcd-snapshot to use config file with flags that are only used with k3s server.

### PR DESCRIPTION
**Pull request by dereknola**
#### Proposed Changes ####
Allows `server` and `etcd-snapshot` to ignore flags in the `config.yaml` that do not match their valid flags list. This does not affect command line arguments.

#### Verification ####
New unit test

Start k3s server with the following `config.yaml`:
```
token: 12345
node-label: myNodeIsFun
etcd-s3: true
```

Run `k3s etcd-snapshot save` and verify it does not fail. You should only see warnings about unknown flags

#### Linked Issues ####
https://github.com/k3s-io/k3s/issues/4449

#### User-Facing Change ####

```release-note

```

### CODE REVIEW

File: pkg/configfilearg/parser.go [hunk 1]
```diff
@@ -44,12 +47,53 @@ func (p *Parser) Parse(args []string) ([]string, error) {
...
+			values = p.stripInvalidFlags(args[1], values)
+		}
 		return append(prefix, append(values, suffix...)...), nil
 	}
 
 	return args, nil
 }
 
+func (p *Parser) stripInvalidFlags(command string, args []string) []string {
+
```
Comment (briandowns) [inline pkg/configfilearg/parser.go:60]: Nit: kill this empty line

Comment (dereknola) [reply pkg/configfilearg/parser.go]: killed

File: pkg/configfilearg/defaultparser.go [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
 	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
 )
 
+var defaultParser = &Parser{
+	After:         []string{"server", "agent", "etcd-snapshot:1"},
+	FlagNames:     []string{"--config", "-c"},
+	EnvName:       version.ProgramUpper + "_CONFIG_FILE",
+	DefaultConfig: "/etc/rancher/" + version.Program + "/config.yaml",
+	ValidFlags:    map[string][]cli.Flag{"server": cmds.ServerFlags, "etcd-snapshot": cmds.EtcdSnapshotFlags},
```
Comment (brandond) [inline pkg/configfilearg/defaultparser.go:15]: do we need to handle `agent` in here too? Not sure if it's necessary but it feels like an omission.

File: pkg/configfilearg/defaultparser_test.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+			args: []string{"k3s", "etcd-snapshot", "save", "--s3=true"},
+
+			want: []string{"k3s", "etcd-snapshot", "save", "--s3=true"},
+		},
+		{
+			name:   "Etcd-snapshot with config with known and unknown flags",
+			args:   []string{"k3s", "etcd-snapshot", "save"},
+			config: "./testdata/defaultdata.yaml",
+			want:   []string{"k3s", "etcd-snapshot", "save", "--etcd-s3=true", "--etcd-s3-bucket=my-backup"},
+		},
```
Comment (brandond) [inline pkg/configfilearg/defaultparser_test.go:51]: Should we test loading agent config as well?

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/configfilearg/defaultparser.go (hunk 2 above)
Comment (dereknola) [reply pkg/configfilearg/defaultparser.go:15]: I can expand it to handle agent, I was focused on server and etcd-snapshot because of the bug

Comment (dereknola) [reply pkg/configfilearg/defaultparser.go:15]: Nothing happens if we have agent, it just returns without trying to skip any arguments.

File: pkg/configfilearg/defaultparser_test.go (hunk 3 above)
Comment (dereknola) [reply pkg/configfilearg/defaultparser_test.go:51]: Added agent subtests

**Review by brandond — APPROVED** [review]

**Review by briandowns — APPROVED** [review]

**Review by brandond — APPROVED** [review]

**Pull request merged**